feat: initial implementation

f0rmiga opened the pull request:
This is the initial implementation of the KubeCF Apps DNS.
It's comprised of a CoreDNS plugin that translates responses from the Service Discovery Controller into DNS A and AAAA answers.

Commit:
feat: add kubecf-tools submodule

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
feat: initial version

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: release body message

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

mook-as commented on .github/workflows/release.yaml:
```diff
@@ -0,0 +1,57 @@
+      env:
+        GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}
+    - name: Build and publish image
+      uses: docker/build-push-action@v1
```
FWIW, they are starting to print deprecation notices with `v1` now. Probably don't need to worry about it yet.

f0rmiga replied:
I'll address this on a follow-up PR.

mook-as commented on README.md:
```diff
@@ -1 +1,8 @@
+
+This is the Apps DNS for KubeCF. It's a CoreDNS plugin that performs service
+discovery using the Service Discovery Controller and responds with discovered
+routes.
```
It would be good to give some examples for the kind of things this resolves (like `go-env.cf.app`, or whatever it should be).

f0rmiga replied:
Fixed with 027350a.

mook-as commented on image/Dockerfile:
```diff
@@ -0,0 +1,36 @@
+COPY svcdiscovery/plugin.cfg plugin.cfg
+RUN go generate
+RUN make
+WORKDIR /build
+COPY svcdiscovery/resolvwriter/ ./resolvwriter
```
```suggestion
COPY svcdiscovery/resolvwriter/ /build/resolvwriter
```
Why do we have the excess `WORKDIR`? Can we get rid of more of them?

Of course, I mentally think of `resolvwriter` as a completely separate thing, and wonder if we could build it in a different stage just to be clear… but that's probably overkill.

f0rmiga replied:
Addressed with a9aeda4.

mook-as commented on svcdiscovery/plugin.cfg:
```diff
@@ -0,0 +1,8 @@
+cache:cache
+loop:loop
+reload:reload
+loadbalance:loadbalance
```
I notice that the ordering here is different from [the default]; as that says the ordering is important, it will probably be a good idea to have some more context on the particular ordering chosen.  Ideally, that would be embedded in a commit message (where just this was changed), because I don't trust things like GitHub comments that don't get easily replicated off their servers.

[the default]: https://github.com/coredns/coredns/blob/v1.7.0/plugin.cfg

f0rmiga replied:
Weird... their comment is less than optimal. I don't really understand what they mean by "feel the effects" - sounds more like something I'd read from a psychology book rather than a software component. As always, it's just easier to go read the source code to understand **how** the ordering here affects the DNS behaviour.

f0rmiga replied:
Fixed with 93faaa3.

mook-as commented on svcdiscovery/resolvwriter/main.go:
```diff
@@ -0,0 +1,84 @@
+/*
```
As `resolvwriter` is a stand-alone binary, please include a separate README for it; at a minimum, we should describe why it exists.  It looks like it could be replaced by a 3 line shell script…

It is unclear to my why this is under `svcdiscovery/…` rather than at the top level (in its own directory).

f0rmiga replied:
While it could be a 3-line shell script, it's on an image that has no shell! :D
I was even writing this in C but felt too complicated for other people to do maintenance in the future.
Regarding where it lives, I'm fine putting it wherever with its own description.

f0rmiga replied:
Addressed with a9aeda4.

mook-as commented on svcdiscovery/resolvwriter/main.go:
```diff
@@ -0,0 +1,84 @@
+	for {
+		var err error
+		ips, err = net.LookupIP(host)
+		if err != nil {
```
It might be clearer to invert the condition and just break early on success (or return early, that works too).  I got confused initially because the `break` was too far away from the `if`… I think I might be getting dumber.

f0rmiga replied:
Whenever I read `if err == nil`, my brain gets a short circuit. In this case, I'm the one getting dumber!

mook-as commented on svcdiscovery/resolvwriter/main.go:
```diff
@@ -0,0 +1,84 @@
+		ips, err = net.LookupIP(host)
+		if err != nil {
+			dnsErr, ok := err.(*net.DNSError)
+			if ok && (dnsErr.Temporary() || dnsErr.IsNotFound) {
```
Retrying on not found is… unusual. I assume this is because it's a Kubernetes service and it'll have a result soon; but we should probably put a comment here or something (or even better, the readme), so that if we reuse this thing we can figure out what it's doing (and why).

f0rmiga replied:
Fair enough. I'll add both the comment and the README.

f0rmiga replied:
Fixed with d72b49d.

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,116 @@
+	ttl       uint32
+}
+
+// Name returns the plugin name.
```
```suggestion
// Name of the plugin.
```
(We _don't_ need to always stick to `… returns` for getters)

Presumably this is to satisfy the `plugin.Handler` interface? If yes, we should document that.

f0rmiga replied:
Fixed with 20607dd.

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,116 @@
+func (sd *ServiceDiscovery) Name() string { return pluginName }
+
+// ServeDNS handles domain name requests for Cloud Foundry App Service Discovery
+// by resolving internal domain names.
```
Something about implementing `plugin.Handler`?

f0rmiga replied:
Fixed with 20607dd.

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,116 @@
+		}
+
+		if len(ips) > 0 {
+			res = &ServiceDiscoveryResponseWriter{
```
Hmm, their sample for a plugin that actually resolves a thing is the [whoami](https://github.com/coredns/coredns/blob/master/plugin/whoami/whoami.go) plugin.  That code makes more sense: if it's got a proper response, just use the `ResponseWriter` to write it out, and don't pass it onto the next plugin.  That would mean we don't need to write the whole `ResponseWriter`.

In our case though we probably want to chain to the next plugin if we would have `NXDOMAIN`ed.

f0rmiga replied:
Fixed with ac0eb20.

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,116 @@
+		ips, err := sd.sdcClient.Discover(qname)
+		if err != nil {
+			sd.log.Error(err)
+			return dns.RcodeServerFailure, err
```
If I understand things correctly, we're return a server failure for DNS calls that we don't cover (say, looking up `www.google.com`) that happens to error out? That seems… odd.  But maybe that's okay, if we expect that to be very rare?

f0rmiga replied:
No. An error would only occur on connection failure (or something else other than not finding google.com).
When the service discovery controller doesn't have any routes to respond, we get a success response, but no IPs.
If you continue reading, you'll see that we only set out a response writer if the list of IPs is not empty. Any other case would just be forwarded to the upstream server (configured on the Corefile on KubeCF).

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,116 @@
+// WriteMsg writes the response for the discovered service.
+func (sdrw *ServiceDiscoveryResponseWriter) WriteMsg(res *dns.Msg) error {
+	answer := make([]dns.RR, 0)
+	qtype := res.Question[0].Qtype
```
What if there was more than one question? (Is that a thing? The structure looks like an array, so maybe?)

f0rmiga replied:
Multiple questions are not a reality. You can look up online and you will repeatedly get the same answer: nobody practices this and it's not well defined how the behaviour should be.

mook-as commented on svcdiscovery/plugin/svcdiscovery/sdcclient.go:
```diff
@@ -0,0 +1,79 @@
+// Discover discovers internal app routes from the Service Discovery Controller
+// and returns the list of IPs from these discovered routes.
+func (sdcc *SDCClient) Discover(domainName string) ([]string, error) {
+	url := sdcc.sdcURLBase + domainName
```
I think I'd be a lot happier if the URL is a proper [URL](https://pkg.go.dev/net/url?tab=doc#URL) instead of a giant string (mostly to have a better chance at escaping things properly). Especially since we actually do that, and just convert it to string too early…

f0rmiga replied:
AFAICT, the way I'm doing it is way more efficient than dealing with a URL object here (or even fmt.Sprintf).
Since the `domainName` string comes from a DNS Question Name, can this ever get wrong?

mook-as commented on svcdiscovery/plugin/svcdiscovery/setup.go:
```diff
@@ -0,0 +1,156 @@
+		var ttl uint32
+		for c.NextBlock() {
+			switch c.Val() {
+			case "tls_ca_path":
```
Most of these are a plain string; they look very repetitive, and I wonder if it would be possible to reduce some of the repetition without making the code too much harder to read.  Also, do we need to check that `sdcPort` is a proper number? Or that we're not getting values that make no sense (since we don't have a `default` block in the `switch`).

f0rmiga replied:
1. +100 on missing `default` statement. I'm usually the one to complain when it is missing.
2. Repetition is not inherently bad. I don't want to spend time trying to simplify this since it's already very simple, just repetitive.
3. I can certainly do validation on the `sdcPort`.

f0rmiga replied:
Added validation to `sdc_port` with b2841b9.

f0rmiga replied:
Added missing `default` case with 5daee74.

Commit:
fix: validation to sdc_port

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: expand README

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: missing default case for switch

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
refactor: resolvwriter

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: comment for retry on resolvwriter

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: interface satisfaction comments

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: plugin order on plugin.cfg

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: plugin ordering and response

The response is built from scratch and doesn't use a custom response
writer. Also, process the svcdiscovery before the forward plugin.

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
feat: add logging to svcdiscovery responses

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

mook-as commented on README.md:
```diff
@@ -1 +1,14 @@
+
+For examples on how to perform container-to-container (c2c) networking and
+service discovering, refer to the following link:
+https://github.com/cloudfoundry-attic/cf-networking-examples/blob/8e3be5faa135746cfd1e05bc618fe18053eab2b5/docs/c2c-with-service-discovery.md
```
```suggestion
service discovering, please refer to [the documentation example](
https://github.com/cloudfoundry-attic/cf-networking-examples/blob/8e3be5faa135746cfd1e05bc618fe18053eab2b5/docs/c2c-with-service-discovery.md).
```

(The raw URL doesn't look nice.)

Commit:
fix: return rcode on write

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

mook-as commented on resolvwriter/main.go:
```diff
@@ -0,0 +1,86 @@
+	}
+}
+
+// resolve resolves the host and returns all IPs.
```
```suggestion
// resolve the host and returns all IP addresses associated with it.
```

mook-as commented on resolvwriter/main.go:
```diff
@@ -0,0 +1,86 @@
+	return ips, nil
+}
+
+// render renders the output file with the provided nameservers. The output is
```
```suggestion
// render the output file with the provided nameservers. The output is
```

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,111 @@
+func (sd *ServiceDiscovery) Name() string { return pluginName }
+
+// ServeDNS handles domain name requests for Cloud Foundry App Service Discovery
+// by resolving internal domain names. It satisfies satisfies
+// plugin.Handler.ServeDNS.
```
```suggestion
// by resolving internal domain names. It satisfies plugin.Handler.ServeDNS.
```

mook-as commented on svcdiscovery/plugin/svcdiscovery/sdcclient.go:
```diff
@@ -0,0 +1,79 @@
+
+// Discover discovers internal app routes from the Service Discovery Controller
+// and returns the list of IPs from these discovered routes.
+func (sdcc *SDCClient) Discover(domainName string) ([]string, error) {
```
This should take a `Context`, so that we can correctly abort the HTTP request or something.  Unfortunately I think that involves [`NewRequestWitheContext`](https://pkg.go.dev/net/http/?tab=doc#NewRequestWithContext).

f0rmiga replied:
Mostly useful for tracing than cancelling in this case. Anyways, this is simple to accomplish here.

f0rmiga replied:
Done in 83c38bb.

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,111 @@
+
+	res := &dns.Msg{}
+	res.SetReply(req)
+	res.Authoritative = true
```
Hmm, I thought we're basically recursing (into a different protocol); are we considered authoritative here?

f0rmiga replied:
Yes, we are authoritative to respond to IPs of discovered services from the Service Discovery Controller. Anything else we recurse.

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,111 @@
+	res.SetReply(req)
+	res.Authoritative = true
+
+	answer := make([]dns.RR, 0)
```
```suggestion
	answer := make([]dns.RR, 0, len(ips))
```
No reason not to allocate ahead of time; since the capacity is separate from the length, it's not an issue if we end up not using some IPs.

f0rmiga replied:
Hmmmm... while the logic doesn't guarantee it, I guess it will always be the case. I thought about it and decided to leave it out, but I'll take it.

mook-as commented on svcdiscovery/plugin/svcdiscovery/plugin.go:
```diff
@@ -0,0 +1,111 @@
+	}
+	res.Answer = answer
+
+	sd.log.Infof("%s: %+v\n", qname, answer)
```
Do we need to log this at `Info`? I'd have thought `Debug` would be a more appropriate level here.

f0rmiga replied:
Yeah... debug makes more sense, I guess.

f0rmiga replied:
Fixed with 6b50c04.

mook-as reviewed (changes requested):
Mostly because we forgot about context.

I'd also like to see tests, or at least an issue filed to add tests.

Commit:
fix: suggestions from code review

Co-authored-by: Mark Yen <3977982+mook-as@users.noreply.github.com>

Commit:
fix: debug level for logging

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
feat: pass context to sdc client

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

Commit:
fix: missing error logging

Signed-off-by: Thulio Ferraz Assis <thulio.assis@suse.com>

mook-as commented on resolvwriter/main.go:
```diff
@@ -0,0 +1,86 @@
+	var ips []net.IP
+	for {
+		var err error
+		ips, err = net.LookupIP(host)
```
Note to future-self: not using a context here is okay, since this thing runs to completion, and the only context we have here is `Background`.

f0rmiga replied:
We could create other contexts, but none would bring any benefit.

mook-as commented on svcdiscovery/plugin/svcdiscovery/sdcclient.go:
```diff
@@ -0,0 +1,84 @@
+// and returns the list of IPs from these discovered routes.
+func (sdcc *SDCClient) Discover(ctx context.Context, domainName string) ([]string, error) {
+	url := sdcc.sdcURLBase + domainName
+	req, err := http.NewRequestWithContext(ctx, "GET", url, nil)
```
```suggestion
	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
```

mook-as reviewed (dismissed):
Only trivial suggestion left.

Commit:
fix: use proper const

Co-authored-by: Mark Yen <3977982+mook-as@users.noreply.github.com>

f0rmiga commented:
This is part of https://github.com/cloudfoundry-incubator/kubecf/issues/776.
